app/frontend/wallet/shelley/shelley-transaction-planner.ts

```diff
@@ -1,315 +1,334 @@
+    if (txPlanArgs.txType === TxType.SEND_ADA && txPlanArgs.donationAmount > 0) {
+      outputs.push({
+        address: getDonationAddress(),
+        coins: txPlanArgs.donationAmount,
+      })
+    }
+    return {
+      outputs,
+      certificates: [],
+      withdrawals: [],
+    }
   }
-  const profitableUtxos = utxos.filter(isUtxoProfitable)
-
-  const inputs = []
 
-  const outputs = address ? [{address, coins, accountAddress}] : []
-  if (donationAmount > 0) {
-    outputs.push({
-      address: getDonationAddress(),
-      coins: donationAmount,
-      accountAddress,
-    })
+  const prepareDelegationTx = (txPlanArgs: DelegateAdaTxPlanArgs): TxPlanDraft => {
+    const certificates: _Certificate[] = []
+    if (!txPlanArgs.isStakingKeyRegistered) {
+      certificates.push({
+        type: CertificateType.STAKING_KEY_REGISTRATION,
+        stakingAddress: txPlanArgs.stakingAddress,
+        poolHash: null,
+      })
+    }
+    if (txPlanArgs.poolHash) {
+      certificates.push({
+        type: CertificateType.DELEGATION,
+        stakingAddress: txPlanArgs.stakingAddress,
+        poolHash: txPlanArgs.poolHash,
+      })
+    }
+    return {
+      outputs: [],
+      certificates,
+      withdrawals: [],
+    }
   }
 
-  const change = {address: changeAddress, coins: 0 as Lovelace, accountAddress}
+  const prepareWithdrawalTx = (txPlanArgs: WithdrawRewardsTxPlanArgs): TxPlanDraft => {
+    const withdrawals: _Withdrawal[] = []
+    withdrawals.push({stakingAddress: txPlanArgs.stakingAddress, rewards: txPlanArgs.rewards})
+    return {
+      outputs: [],
+      certificates: [],
+      withdrawals,
+    }
+  }
 
-  for (let i = 0; i < profitableUtxos.length; i++) {
-    inputs.push(profitableUtxos[i])
-    const plan = computeTxPlan(inputs, outputs, change, certs, withdrawals)
-    if (plan) return plan
+  switch (txPlanArgs.txType) {
+    case TxType.SEND_ADA:
+      return prepareSendAdaTx(txPlanArgs)
+    case TxType.DELEGATE:
+      return prepareDelegationTx(txPlanArgs)
+    case TxType.WITHDRAW:
+      return prepareWithdrawalTx(txPlanArgs)
+    case TxType.CONVERT_LEGACY:
+      return prepareSendAdaTx(txPlanArgs)
+    default:
+      throw NamedError('InvalidTxPlanType')
   }
+}
 
+export const selectMinimalTxPlan = (
+  utxos: Array<UTxO>,
+  changeAddress: _Address,
+  txPlanArgs: TxPlanArgs
+): TxPlanResult => {
+  const inputs: _Input[] = []
+  const {outputs, certificates, withdrawals} = prepareTxPlanDraft(txPlanArgs)
+  const change: _Output = {address: changeAddress, coins: 0 as Lovelace}
+
+  // TODO: refactor this when implementing multi assets
+  for (const utxo of utxos) {
+    inputs.push(utxo)
+    try {
+      const plan = computeTxPlan(inputs, outputs, change, certificates, withdrawals)
+      validateTxPlan(plan)
+      return {
+        type: TxPlanResultType.SUCCESS,
+        txPlan: plan,
+      }
+    } catch (e) {
+      if (inputs.length === utxos.length) {
```

Comment:
wtf are we doing here?
This at minimum should be just after the cycle though a deeper question is whether compute/validate tx plan needs to throw ...

Reply:
There is a comment above, that we should refactor this, I just made it so it works for now.. :D. Regarding the throwing in computing tx plan, we need information about what caused the plan to not be created. So I intentionally made it to throw errors.

Reply:
I suppose we can drop the the throwing in the plan computation and validate it in the `validateTxPlan`. in that case tho, we would do a lot of stuff twice, (like computing the size of tx in bytes...) but I guess it would separate the concert of tx plan validation... :thinking: